File: fluo-rts.R
#
require(readr)
require(nnls)

#
source_eem_data <- readr::read_csv("source_eems.csv", name_repair = "minimal")
unknown_eem_data <- readr::read_csv("unknown_eems.csv")
source_labels <- colnames(source_eem_data)

#
sources <- unique(source_labels)
K <- length(sources)
N <- length(source_labels)
G <- matrix(0, nrow = K, ncol = N)
for(k in 1:K){
  for(n in 1:N){
    if(source_labels[n] == sources[k]){
      G[k, n] <- 1
    }
  }
}

#
source_contributions <- matrix(0, nrow = ncol(unknown_eem_data), ncol = K)
for(j in 1:ncol(unknown_eem_data)){
  #
  y <- unknown_eem_data[, j]
  X <- as.matrix(source_eem_data)
    
  #
  incl_rows <- apply(X, 1, function(x){!any(is.na(x))}) & !is.na(y)
  X <- X[incl_rows, ]
  y <- y[incl_rows]
  
  #
  coefs <- nnls::nnls(X, y)$x
  
  #
  source_contributions[j, ] <- G%*%coefs
}

#
source_contributions <- as.data.frame(source_contributions)
colnames(source_contributions) <- sources
source_contributions <- cbind(eem_name = colnames(unknown_eem_data), source_contributions)
readr::write_csv(source_contributions, "source_contributions.csv")

File: fluo-rts.py
from scipy.optimize import nnls
import numpy as np
import pandas as pd
import re

#
source_eem_data = pd.read_csv("source_eems.csv")
source_labels = source_eem_data.columns.values
source_labels = np.array([re.sub("[.].*", "", x) for x in source_labels])

#
unknown_eem_data = pd.read_csv("unknown_eems.csv")
unknown_labels = unknown_eem_data.columns.values

#
sources = np.unique(source_labels)
K = sources.size
N = source_labels.size
G = np.zeros([K, N])
for k in range(K):
  for n in range(N):
    if source_labels[n] == sources[k]:
      G[k, n] = 1

#
source_contributions = np.zeros([unknown_labels.size, K])
for j in range(unknown_labels.size):
  #
  b = unknown_eem_data.values[:, j]
  A = source_eem_data.values

  #
  na_b = np.isnan(b)
  na_A = np.isnan(A).any(axis=1)

  #
  b = b[~na_b * ~na_A]
  A = A[~na_b * ~na_A, :]
  
  #
  coefs = nnls(A, b)[0]
  source_contributions[j, :] = np.matmul(G, coefs)

#
pd.DataFrame(data=source_contributions, index=unknown_labels, columns=sources).to_csv("source_contributions.csv")
